### PR TITLE
Feat/refresh access token

### PR DESCRIPTION
**Pull request by pavel-sobchenko**
https://github.com/users/rasulovdev/projects/3/views/1?pane=issue&itemId=30474977

add refresh jwt

### CODE REVIEW

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+): Observable<HttpEvent<unknown>> => {
+  const apiService = inject(AuthService)
+  const tokenStorage = inject(TokensStorageService)
+  const router = inject(Router)
+  const store = inject(Store)
+  const toastr = inject(ToastrService)
+
+  return next(req).pipe(
+    catchError((error: HttpErrorResponse) => {
+      if (error.status === 401 && !req.url.includes('new')) {
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: `new` is not specific enough

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        if (tokenStorage.get()?.accessToken) {
+          const token = tokenStorage.get()?.refreshToken
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: Maybe check `refreshToken` if you need **that** further?

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+          return apiService.refreshToken(token).pipe(
+            switchMap((tokens: Tokens) => {
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: Do not specify easily inferrable types

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                  Authorization: `Bearer ${tokens.accessToken}`
+                }
+              })
+
+              return next(req)
+            }),
+            catchError((err: ErrorResponse) => {
+              if (err.statusCode === 403) {
+                store.dispatch(new LogOut())
+                void router.navigate(['auth/login'])
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: You can automate this logic with NGXS's `ofActionDispatched` operator or similar

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  return next(req).pipe(
+    catchError((error: HttpErrorResponse) => {
+      if (error.status === 401 && !req.url.includes('new')) {
+        if (tokenStorage.get()?.accessToken) {
+          const token = tokenStorage.get()?.refreshToken
+
+          return apiService.refreshToken(token).pipe(
+            switchMap((tokens: Tokens) => {
+              tokenStorage.set(tokens)
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: You introduces data mismatch with this expression. Check both state and localstorage

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+              req = req.clone({
+                setHeaders: {
+                  Authorization: `Bearer ${tokens.accessToken}`
+                }
+              })
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: I don't think you need this

**Review by rasulovdev — CHANGES_REQUESTED** [review]

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        return nextFn(req)
+      }),
+      catchError((err: HttpErrorResponse) => {
+        if (err.status === 403) {
+          store.dispatch(new LogOut())
+        }
+        toastr.error((err.error as ErrorResponse).message, err.statusText)
+
+        return throwError(() => err)
+      })
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: `handle401Error` function should not handle anything other than 401 error

File: src/app/core/http/interceptors/utils.ts
Comment (rasulovdev) [inline src/app/core/http/interceptors/utils.ts:1]: I don't think you need this

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts (hunk 7 above)
Comment (rasulovdev) [reply src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: This is not correct fix. You needed to only extract logic for handling 401 error, not everything. Now this new method does everything again, just in another place

File: src/app/layout/pages/main-dashboard-page/main-dashboard-page.component.ts [hunk 8]
```diff
@@ -20,10 +20,23 @@ export class MainDashboardPageComponent {
     this._store.select(AuthStateSelectors.isAuthenticated)
   )
 
-  constructor(private readonly _store: Store) {
+  private readonly _isLoggedOut = toSignal(
```
Comment (rasulovdev) [inline src/app/layout/pages/main-dashboard-page/main-dashboard-page.component.ts]: when you name a variable with `is`, it implies the value of the variable will be boolean. This is not the case here. I will not get boolean (true/false) when I call `this._isLoggedOut()`, right?

**Review by rasulovdev — CHANGES_REQUESTED** [review]

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    return apiService.refreshToken(token).pipe(
+      switchMap(tokens => {
+        store.dispatch(new SetTokens(tokens))
+
+        return nextFn(req)
+      }),
+      catchError((err: HttpErrorResponse) => {
+        return throwError(() => err)
+      })
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: This is doing nothing

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    )
+  }
+
+  function handle403Error(
+    request: HttpRequest<unknown>,
+    nextFn: HttpHandlerFn
+  ): Observable<HttpEvent<unknown>> {
+    store.dispatch(new LogOut())
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: Why would you log out a user, who is legitimately logged in, but just accessed a resource they don't permission to? They still have permissions to view a lot of other things in your app

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+      }),
+      catchError((err: HttpErrorResponse) => {
+        return throwError(() => err)
+      })
+    )
+  }
+
+  function handle403Error(
+    request: HttpRequest<unknown>,
+    nextFn: HttpHandlerFn
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: this is not being used

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  ): Observable<HttpEvent<unknown>> {
+    store.dispatch(new LogOut())
+    toastr.error(
+      'You do not have permission to perform this action.',
+      'Forbidden'
+    )
+
+    return throwError(() => new HttpErrorResponse({ ...request, status: 403 }))
```
Comment (rasulovdev) [inline src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: Same thing is happening on line 30. The job of this function is to just show a toast, that's it

File: src/app/core/http/interceptors/utils.ts
Comment (rasulovdev) [inline src/app/core/http/interceptors/utils.ts:1]: You don't need this anymnore, right?

**Review by rasulovdev — CHANGES_REQUESTED** [review]

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts (hunk 9 above)
Comment (pavel-sobchenko) [reply src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: Generally, you are right. In this case we can avoid this construction.

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts (hunk 10 above)
Comment (pavel-sobchenko) [reply src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: I didn't think about the case with view possibility. I thought, if a user doesn't have permission to make a web API call it means he should be logged out automatically.

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts (hunk 11 above)
Comment (pavel-sobchenko) [reply src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: fixed

File: src/app/core/http/interceptors/refresh-jwt.interceptor.ts (hunk 12 above)
Comment (pavel-sobchenko) [reply src/app/core/http/interceptors/refresh-jwt.interceptor.ts]: fixed

File: src/app/core/http/interceptors/utils.ts
Comment (pavel-sobchenko) [reply src/app/core/http/interceptors/utils.ts:1]: yes. removed

**Review by rasulovdev — APPROVED** [review]

**Pull request merged**